proposals/accepted/047-standard-library-reform.rst

```diff
@@ -0,0 +1,559 @@
+
+However, this doesn't go far enough to address the standard library --- language implementation coupling problem.
+Both libraries still live in the compiler repo and are still released in tandem with the compiler.
+``core`` also contains numerous definitions that, while perfectly portable, have nothing to do with interfacing the compiler internals.
+(Think e.g. the equivalents of things like ``Functor`` and ``Monoid`` for us, perfectly portable across compilation targets, but also implementation-agnostic.)
+
+Rust's ``cap-std``
+~~~~~~~~~~~~~~~~~~
+
+`cap-std <https://github.com/bytecodealliance/cap-std>`_ is a Rust library exploring what ergonomic IO interfaces for WASI system calls in a high level language should look like.
+On one hand, it is great, and we should borrow from it heavily.
+On the other hand, we should surpass it in not needing to be something on top of the "regular" standard library which ordinarily exposes more Unixy things than is appropriate.
+
+Rust's ``std`` and WASI
+~~~~~~~~~~~~~~~~~~~~~~~
+
+While the best experience comes from using ``cap-std`` as described above, Rust's ``std`` still makes sure to avoid indirecting through ``wasi-libc`` wherever possible.
+`This PR <https://github.com/rust-lang/rust/pull/63676>`_ made that change, using the ``wasi`` library (Rust bindings to WASI system calls) directly.
+This is what we should emulate in order to provide a top-tier programming environment for greenfield WebAssembly applications in Haskell.
+
+Prior attempts at splitting ``base``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+There have been attempts to split ``base`` before, but they attempted to get everything done at once, setting a dangerously high bar for success.
+This approach here, by contrast, first and foremost seeks to avoid those difficulties and find a sustainable, suitably low-risk approach.
+It is much more concerned with how we safely approach these issues than what the exact outcome looks like.
+
+Technical Content
+-----------------
+
+Here is a plan to solve these issues.
+
+**Step 1A**: Task the CLC with defining new standard libraries
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Based on the conversation in `CLC Issue #015`_, ``base`` is exposing too much stuff, yet trying to limit what is exposed would be a big breaking change.
+
+The solution is to reach for another layer of indirection.
+The CLC should be tasked with devising new standard library interfaces, which would initially be implemented by reexporting modules from ``base``.
+
+The new library interfaces should be carefully designed in and of themselves to tackle many, but not all, of the issues above:
+
+- They should be designed *not* to break every release.
+  Even though the underlying ``base`` from which modules are reexported would continue to have its regular problematic major version bumps, the portion reexported should have very infrequent breaking changes.
+
+  This fixes **Problem 1**.
```

Comment:
I don't think it does, because nobody will be using it. You'll also have to solve the buy-in problem and arrange for most important Haskell projects to be using this new library. If it's called `standard`, for instance, then `cabal init` and `stack init` will have to be emitting it as the default dependency for a while before this actually happens. We'd also have to make public noise about the change in direction. But the plan doesn't say anything about how to get build tools or developers in general on board.